Rename pipelines

amn41 opened the pull request:
**Proposed changes**:
- fixes #1582 
- renames standard templates to reflect what they *do* rather than what libraries they use
- updates docs
- maintains backwards compatibility, old template names are still permitted but a warning is printed


**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

wochinge commented on docs/components.rst:
```diff
@@ -4,6 +4,12 @@
 
+.. note::
+   For clarity, we have renamed the pre-defined pipelines to reflect what they *do*
+   rather than which libraries they use as of Rasa NLU 0.13. The ``tensorflow_embedding`` 
```
```suggestion
   rather than which libraries they use as of Rasa NLU 0.14. The ``tensorflow_embedding`` 
```

wochinge commented on rasa_nlu/config.py:
```diff
@@ -125,6 +125,15 @@ def __init__(self, configuration_values=None):
+                new_names = {"spacy_sklearn": "pretrained_embeddings",
+                             "tensorflow_embedding": "supervised_embeddings"}
+                if template_name in new_names:
+                    logger.warn("You have specified the pipeline template {} "
```
```suggestion
                    logger.warning("You have specified the pipeline template {} "
```

wochinge replied:
This suggestion will probably break the indentation of the lines beyond.

wochinge commented on docs/choosing_pipeline.rst:
```diff
@@ -201,14 +201,14 @@ a full list of components. For example, these two configurations are equivalent:
+Pretrained_embeddings
+~~~~~~~~~~~~~~~~~~~~~
 
 To use spacy as a template:
```
maybe: `To use a pipeline which uses pretrained embeddings`?

wochinge commented on docs/languages.rst:
```diff
@@ -5,27 +5,28 @@ Language Support
 	
 Another way is to define a custom configuration by listing all components you would like your pipeline to use.
-The tensorflow pipeline supports any language that can be tokenized. The default is to use a simple 
+The supervised embeddings pipeline supports any language that can be tokenized. The default is to use a simple 
```
```suggestion
The ``supervised embeddings`` pipeline supports any language that can be tokenized. The default is to use a simple 
```

wochinge commented on docs/languages.rst:
```diff
@@ -75,23 +76,23 @@ Once the training is finished, you can test your model's Norwegian language skil
 ------------------------
 
-With the spaCy backend you can now load fastText vectors, which are available 
+With the pretrained embeddings pipline you can also load fastText vectors, which are available 
```
```suggestion
With the ``pretrained embeddings`` pipeline you can also load fastText vectors, which are available 
```

wochinge commented on rasa_nlu/registry.py:
```diff
@@ -67,6 +67,15 @@
 # necessary components in the right order. They also implement
 # the preexisting `backends`.
 registered_pipeline_templates = {
+    "pretrained_embeddings": [
```
I think we should  remove the old templates and define a mapping from the old template names to the new ones in `pipeline_template` or in `config.py`, but that's open for discussion

amn41 replied:
I agree that would be cleaner, but since we will remove these anyway in next major release I think it's not worth making the code more elegant

wochinge commented:
@amn41 Is this ready or are you still working on it? Did you see bullet points in my comment above?

tmbo commented:
@amn41 I did not have a lot of time to work on this (only made a couple small changes) as there are more important things to do for me atm. If you want you can finish this is up, otherwise it is probably not going to get done for a while.

amn41 commented:
thanks @wochinge ! can you please have another look? I addressed all of your comments. 

Regarding renaming the docker builds, I think we have to wait until this is merged bc docker hub builds from master. unless @tmbo you know of a workaround?

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/5114084/53229485-95e66c80-3684-11e9-8fda-0078d827cf71.png">

tmbo commented:
Yes I think it is fine to do that once merged to master

amn41 commented:
@wochinge can you have a look today?

wochinge commented on docker/Dockerfile_pretrained_embeddings_spacy_de:
```diff
@@ -0,0 +1,34 @@
+
+EXPOSE 5000
+
+ENTRYPOINT ["./entrypoint.sh"]
```
Both images differ regarding the language model. However, I did not find a clean way to reuse the images without having a certain order of docker builds. But could we maybe add the model download / link to the entrypoint so that users can specify it with environment variables?

amn41 replied:
by 'reuse' you mean to use one as a base at build time? I think we can optimize for that later. I think the dockerfiles should be 'batteries included' so that users don't have to first run an entrypoint command to download & install the language model (which can take a long time). lmk if I misunderstood

wochinge replied:
> by 'reuse' you mean to use one as a base at build time?

Yes, exactly. But as far as I've seen this is only possible if you build a base image first, and then extend this image. 

> I think the dockerfiles should be 'batteries included'
Makes sense :+1:

wochinge commented on docs/choosing_pipeline.rst:
```diff
@@ -203,14 +203,14 @@ a full list of components. For example, these two configurations are equivalent:
-spacy_sklearn
-~~~~~~~~~~~~~
+pretrained_embeddings_spacy
+~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

wochinge commented on docs/choosing_pipeline.rst:
```diff
@@ -267,16 +267,16 @@ mitie
     :language: yaml
 
-mitie_sklearn
+mitie_2
```
can we explain the difference between mitie 1 and 2 in a few words?

wochinge commented on docs/docker.rst:
```diff
@@ -13,8 +13,8 @@ Images
 
-- ``spacy``: If you use the :ref:`section_spacy_pipeline` pipeline
-- ``tensorflow``: If you use the :ref:`section_tensorflow_embedding_pipeline`
+- ``spacy``: If you use the :ref:`section_pretrained_embeddings_spacy_pipeline` pipeline
```
this is then `spacy_en` and `spacy_de` right?

amn41 replied:
To be consistent, we should rename these to be consistent with the pipeline names, so would be `pretrained_embeddings_spacy_{en, de}`, `supervised_embeddings`, etc.
However these are quite long for tag names, so we could stick with short names like `spacy_en`, `tensorflow`, etc

wochinge commented on docs/installation.rst:
```diff
@@ -88,7 +93,9 @@ memory to run, but will somewhat reduce intent classification performance.
 
-To use the ``tensorflow_embedding`` pipeline you will need to install tensorflow as well as the scikit-learn and sklearn-crfsuite libraries. To do this, run the following command:
+To use the ``supervised_embeddings`` pipeline you will need to install
+tensorflow and, for entity recognition, the sklearn-crfsuite library.
```
```suggestion
TensorFlow and, for entity recognition, the sklearn-crfsuite library.
```

wochinge commented on docs/installation.rst:
```diff
@@ -107,55 +116,42 @@ First, run
+	
+Let us know how you are getting on! If you have any questions
+about the starter-pack or using Rasa NLU in general, post your questions on
+`Rasa Community Forum <https://forum.rasa.com>`_!
```
```suggestion
the `Rasa Community Forum <https://forum.rasa.com>`_!
```

wochinge commented on docs/languages.rst:
```diff
@@ -6,28 +6,31 @@
-    pipeline: "tensorflow_embedding"
-
+    pipeline: "supervised_embeddings"
+	
```
```suggestion

```

wochinge commented on docs/languages.rst:
```diff
@@ -77,23 +80,23 @@ Once the training is finished, you can test your model's Norwegian language skil
+pretrained-embeddings	supported languages
+=====================   =================================
+spacy            	english (``en``),
+               		german (``de``),
```
I get a warning here because of an inconsistent mix of tabs and spaces

wochinge commented on rasa_nlu/config.py:
```diff
@@ -111,6 +111,15 @@ def __init__(self, configuration_values=None):
+                             "tensorflow_embedding": "supervised_embeddings"}
+                if template_name in new_names:
+                    logger.warning("You have specified the pipeline template "
+                                   "{} which has been renamed to {} ."
```
```suggestion
                                   "'{}' which has been renamed to '{}'."
```

wochinge commented on sample_configs/config_supervised_embeddings_duckling.yml:
```diff
@@ -0,0 +1,7 @@
+- name: "intent_featurizer_count_vectors"
+- name: "intent_classifier_tensorflow_embedding"
+- name: "ner_duckling_http"
+  url: "http://duckling:8000"
```
Should we use env variables for the duckling url?

amn41 replied:
possibly, but that's for another issue

tmbo commented on rasa_nlu/config.py:
```diff
@@ -111,6 +111,15 @@ def __init__(self, configuration_values=None):
             if pipeline:
+                new_names = {"spacy_sklearn": "pretrained_embeddings_spacy",
+                             "tensorflow_embedding": "supervised_embeddings"}
+                if template_name in new_names:
```
do we not also need to set the `template_name` from e.g. `"spacy_sklearn"` to `pretrained_embeddings_spacy`?

amn41 replied:
yeah you're right, better to throw a warning and replace the name than to throw an error and quit.
